vault/resource_token.go

```diff
@@ -0,0 +1,301 @@
+
+	if role != "" {
+		log.Printf("[DEBUG] Creating token with role %q", role)
+		resp, err = client.Auth().Token().CreateWithRole(createRequest, role)
+		if err != nil {
+			return fmt.Errorf("error creating token with role %q: %s", role, err)
+		}
+
+		log.Printf("[DEBUG] Created token %q with role %q", resp.Auth.Accessor, role)
+	} else {
+		log.Printf("[DEBUG] Creating token")
+		resp, err = client.Auth().Token().Create(createRequest)
+		if err != nil {
+			return fmt.Errorf("error creating token: %s", err)
+		}
+
+		log.Printf("[DEBUG] Created token %q", resp.Auth.Accessor)
+	}
+
+	d.Set("lease_duration", resp.Auth.LeaseDuration)
+	d.Set("lease_started", time.Now().Format(time.RFC3339))
+	d.Set("client_token", resp.Auth.ClientToken)
+
+	d.SetId(resp.Auth.Accessor)
+
+	return tokenRead(d, meta)
+}
+
+func tokenRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	id := d.Id()
+
+	log.Printf("[DEBUG] Reading token %q", id)
+	resp, err := client.Auth().Token().LookupAccessor(id)
+	if err != nil {
+		log.Printf("[WARN] Token not found, removing from state")
+		d.SetId("")
+		return nil
+	}
+
+	log.Printf("[DEBUG] Read token %q", id)
+
+	if tokenCheckLease(d, client) {
+		log.Printf("[DEBUG] Lease for %q expiring soon, renewing", d.Id())
+		renewed, err := client.Auth().Token().Renew(d.Get("client_token").(string), d.Get("lease_duration").(int))
+		if err != nil {
+			log.Printf("[DEBUG] Error renewing token %q, bailing", d.Id())
+		} else {
+			resp = renewed
+			d.Set("lease_duration", resp.Data["lease_duration"])
+			d.Set("lease_started", time.Now().Format(time.RFC3339))
+			d.Set("client_token", resp.Auth.ClientToken)
+
+			d.SetId(resp.Auth.Accessor)
+		}
+	}
+
+	iPolicies := resp.Data["policies"].([]interface{})
+	policies := make([]string, 0, len(iPolicies))
+	for _, iPolicy := range iPolicies {
+		if iPolicy == "default" {
+			continue
+		}
+
+		policies = append(policies, iPolicy.(string))
+	}
+
+	d.Set("policies", policies)
+	d.Set("no_parent", fmt.Sprintf("%v", resp.Data["orphan"]))
+	d.Set("renewable", fmt.Sprintf("%v", resp.Data["renewable"]))
+	d.Set("display_name", strings.TrimPrefix(resp.Data["display_name"].(string), "token-"))
+	d.Set("num_uses", resp.Data["num_uses"])
+	d.Set("accessor", resp.Data["accessor"])
+
+	return nil
+}
+
+func tokenDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	token := d.Id()
+
+	log.Printf("[DEBUG] Deleting token %q", token)
+	err := client.Auth().Token().RevokeAccessor(token)
+	if err != nil {
+		return fmt.Errorf("error deleting token %q: %s", token, err)
+	}
+	log.Printf("[DEBUG] Deleted token %q", token)
+
+	return nil
+}
+
+func tokenExists(d *schema.ResourceData, meta interface{}) (bool, error) {
+	client := meta.(*api.Client)
+	accessor := d.Id()
+
+	log.Printf("[DEBUG] Checking if token %q exists", accessor)
+	resp, err := client.Auth().Token().LookupAccessor(accessor)
+	if err != nil {
+		return false, nil
```

Comment:
Could the error be logged here for debugging?

Reply:
fixed